
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../register/">
      
      
        <link rel="next" href="../omp/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.34">
    
    
      
        <title>Call Spots - Coppafish Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.35f28582.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="https://unpkg.com/katex@0/dist/katex.min.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#call-spots" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Coppafish Documentation" class="md-header__button md-logo" aria-label="Coppafish Documentation" data-md-component="logo">
      
  <img src="../images/logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Coppafish Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Call Spots
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/reillytilbury/coppafish" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Getting started

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../basic_usage/" class="md-tabs__link">
          
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../overview/" class="md-tabs__link">
        
  
    
  
  Overview

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../stitch/" class="md-tabs__link">
          
  
  Method

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../troubleshoot/" class="md-tabs__link">
        
  
    
  
  Troubleshoot

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../glossary/" class="md-tabs__link">
        
  
    
  
  Glossary

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../contributing/" class="md-tabs__link">
        
  
    
  
  Contributing

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Coppafish Documentation" class="md-nav__button md-logo" aria-label="Coppafish Documentation" data-md-component="logo">
      
  <img src="../images/logo.svg" alt="logo">

    </a>
    Coppafish Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/reillytilbury/coppafish" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Getting started
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Usage
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../basic_usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Basic Usage
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../advanced_usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Advanced Usage
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../overview/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Method
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Method
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../stitch/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Stitch
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../register/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Registration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Call Spots
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Call Spots
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#algorithm-breakdown" class="md-nav__link">
    <span class="md-ellipsis">
      Algorithm Breakdown
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Algorithm Breakdown">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#0-preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      0: Preprocessing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1-initial-gene-assignment" class="md-nav__link">
    <span class="md-ellipsis">
      1: Initial Gene Assignment
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1: Initial Gene Assignment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dye-probabilities" class="md-nav__link">
    <span class="md-ellipsis">
      Dye Probabilities
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene-probabilities" class="md-nav__link">
    <span class="md-ellipsis">
      Gene Probabilities
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-bleed-matrix-calculation" class="md-nav__link">
    <span class="md-ellipsis">
      2: Bleed Matrix Calculation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-free-bled-code-estimation" class="md-nav__link">
    <span class="md-ellipsis">
      3: Free Bled Code Estimation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-round-and-channel-normalisation" class="md-nav__link">
    <span class="md-ellipsis">
      4: Round and Channel Normalisation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-tile-normalisation" class="md-nav__link">
    <span class="md-ellipsis">
      5: Tile Normalisation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-and-7-application-of-scales-computation-of-final-scores-and-bleed-matrix" class="md-nav__link">
    <span class="md-ellipsis">
      6 and 7: Application of Scales, Computation of Final Scores and Bleed Matrix
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#diagnostics" class="md-nav__link">
    <span class="md-ellipsis">
      Diagnostics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Diagnostics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#view-scaling-and-bg-removal" class="md-nav__link">
    <span class="md-ellipsis">
      View Scaling And BG Removal
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#view-bleed-matrix" class="md-nav__link">
    <span class="md-ellipsis">
      View Bleed Matrix
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#view-free-and-constrained-bled-codes" class="md-nav__link">
    <span class="md-ellipsis">
      View Free And Constrained Bled Codes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#view-target-regression" class="md-nav__link">
    <span class="md-ellipsis">
      View Target Regression
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#view-tile-scale-regression" class="md-nav__link">
    <span class="md-ellipsis">
      View Tile Scale Regression
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../omp/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    OMP
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../troubleshoot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Troubleshoot
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../glossary/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Glossary
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../contributing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Contributing
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#algorithm-breakdown" class="md-nav__link">
    <span class="md-ellipsis">
      Algorithm Breakdown
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Algorithm Breakdown">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#0-preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      0: Preprocessing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1-initial-gene-assignment" class="md-nav__link">
    <span class="md-ellipsis">
      1: Initial Gene Assignment
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1: Initial Gene Assignment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dye-probabilities" class="md-nav__link">
    <span class="md-ellipsis">
      Dye Probabilities
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene-probabilities" class="md-nav__link">
    <span class="md-ellipsis">
      Gene Probabilities
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-bleed-matrix-calculation" class="md-nav__link">
    <span class="md-ellipsis">
      2: Bleed Matrix Calculation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-free-bled-code-estimation" class="md-nav__link">
    <span class="md-ellipsis">
      3: Free Bled Code Estimation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-round-and-channel-normalisation" class="md-nav__link">
    <span class="md-ellipsis">
      4: Round and Channel Normalisation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-tile-normalisation" class="md-nav__link">
    <span class="md-ellipsis">
      5: Tile Normalisation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-and-7-application-of-scales-computation-of-final-scores-and-bleed-matrix" class="md-nav__link">
    <span class="md-ellipsis">
      6 and 7: Application of Scales, Computation of Final Scores and Bleed Matrix
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#diagnostics" class="md-nav__link">
    <span class="md-ellipsis">
      Diagnostics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Diagnostics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#view-scaling-and-bg-removal" class="md-nav__link">
    <span class="md-ellipsis">
      View Scaling And BG Removal
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#view-bleed-matrix" class="md-nav__link">
    <span class="md-ellipsis">
      View Bleed Matrix
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#view-free-and-constrained-bled-codes" class="md-nav__link">
    <span class="md-ellipsis">
      View Free And Constrained Bled Codes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#view-target-regression" class="md-nav__link">
    <span class="md-ellipsis">
      View Target Regression
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#view-tile-scale-regression" class="md-nav__link">
    <span class="md-ellipsis">
      View Tile Scale Regression
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="call-spots">Call Spots</h1>
<p>The Call Reference Spots section of the pipeline is a method of gene calling which runs quickly on a small set of spots (<span class="arithmatex">\(\approx\)</span> 50, 000 per tile) of the anchor image. Initially, this was our final mode of gene calling, but has since been superseded by OMP, which differs from Call Spots in that it runs on several more pixels, regardless of whether they have been detected as a spot.</p>
<p>Despite this, the call spots section is still a crucial part of the pipeline as it estimates several important parameters used in the OMP section.</p>
<p>Some of the most important exported parameters of this section are:</p>
<ul>
<li>
<p><strong>Colour Normalisation Factor</strong> <span class="arithmatex">\(\mathbf{A}\)</span>:  <span class="arithmatex">\((n_{\text{t}} \times n_{\text{r}} \times n_{\text{c}})\)</span> array which multiplies the colours to minimise any systematic brightness variability between different tiles, rounds and channels and maximise spectral separation of dyes,</p>
</li>
<li>
<p><strong>Bleed Matrix</strong> <span class="arithmatex">\(\mathbf{B}\)</span>: <span class="arithmatex">\((n_{\text{d}} \times n_{\text{c}})\)</span> array of the typical channel spectrum of each dye.</p>
</li>
<li>
<p><strong>Bled Codes</strong> <span class="arithmatex">\(\mathbf{K}\)</span>: <span class="arithmatex">\((n_{\text{g}} \times n_{\text{r}} \times n_{\text{c}})\)</span> array of the expected colour spectrum for each gene.</p>
</li>
</ul>
<h2 id="algorithm-breakdown">Algorithm Breakdown</h2>
<figure>
  <img alt="Image title" src="../images/algorithm/call_spots/Call%20Spots%20Flowchart.png" />
    <span> Algorithm Flowchart showing how all variables and steps of the pipeline are related.</span>
</figure>
<p>The inputs to the algorithm are:</p>
<ul>
<li>
<p>Raw spot colours <span class="arithmatex">\(F_{src}\)</span> for all spots <span class="arithmatex">\(s\)</span> (defined as local maxima of round <span class="arithmatex">\(r_{\text{anchor}}\)</span>, channel <span class="arithmatex">\(c_{\text{anchor}}\)</span>) and the tile <span class="arithmatex">\(t(s)\)</span> they belong to.</p>
</li>
<li>
<p>A list of genes <span class="arithmatex">\(g\)</span> and their associated dye codes <span class="arithmatex">\(d(g, r)\)</span> for each round <span class="arithmatex">\(r\)</span>. These codes were generated by the Reed-Solomon Algorithm which should minimise the number of overlap between codes.</p>
</li>
<li>
<p>A raw bleed matrix <span class="arithmatex">\(\mathbf{B_{\textrm{raw}}}\)</span> of shape <span class="arithmatex">\((n_{\text{dyes}} \times n_{\text{c}})\)</span> obtained from images of free-floating drops of each dye.</p>
</li>
</ul>
<h3 id="0-preprocessing">0: Preprocessing</h3>
<p>The purpose of this step is to approximately equalise the brightness of different tiles, rounds and channels and to remove any background which is constant across rounds from each spot.</p>
<p>We transform the raw spot colours <span class="arithmatex">\(F_{src}\)</span> as follows:</p>
<div class="arithmatex">\[F_{src} \mapsto \tilde{A}_{t(s)rc}F_{src} - G_{sc}.\]</div>
<p>In the formula above:</p>
<ul>
<li>The <em>initial normalisation factor</em> <span class="arithmatex">\(\tilde{A}_{trc}\)</span> is defined as</li>
</ul>
<div class="arithmatex">\[
\tilde{A}_{trc} = \dfrac{1}{\text{Percentile}_s(F_{src}, 95)}
\]</div>
<p>for all spots <span class="arithmatex">\(s\)</span> in tile <span class="arithmatex">\(t\)</span>. This is a good estimate of the scaling factor needed to make the brightest spots in each tile, round and channel have the same intensity.</p>
<ul>
<li>The <em>ground level</em> <span class="arithmatex">\(G_{sc}\)</span> is defined as</li>
</ul>
<div class="arithmatex">\[
G_{sc} = \text{Percentile}_r(\tilde{A}_{t(s)rc}F_{src}, 25).
\]</div>
<p>For 7 rounds, this is the brightness of the second dimmest round of the scaled spot colours in channel <span class="arithmatex">\(c\)</span>. This is a good estimate of the constant signal in channel <span class="arithmatex">\(c\)</span> across all rounds, which we want to remove.</p>
<!-- TODO: Add an image of a spot before anything, after scaling, then after removing background -->

<h3 id="1-initial-gene-assignment">1: Initial Gene Assignment</h3>
<p>The purpose of this step is to provide some preliminary gene assignments that will allow us to estimate the bleed matrix and the bled codes. We will work extensively with the bleed matrix in these calculations, but bear in mind that this is the raw bleed matrix <span class="arithmatex">\(\mathbf{B_{\textrm{raw}}}\)</span>.</p>
<p>We'd like to define a probability that spot <span class="arithmatex">\(s\)</span> (fluorescence <span class="arithmatex">\(F_{src}\)</span>) comes from gene <span class="arithmatex">\(g\)</span>, and we'd like this probability to have the following properties:</p>
<ol>
<li>
<p>The probability of round <span class="arithmatex">\(r\)</span> being assigned to dye <span class="arithmatex">\(d\)</span> should be invariant to changes in the overall brightness of <span class="arithmatex">\(\mathbf{F_{sr}}\)</span>,</p>
</li>
<li>
<p>the probabilities of each round <span class="arithmatex">\(r\)</span> should be independent.</p>
</li>
</ol>
<details class="note">
<summary>Why these properties?</summary>
<p>Property 1 is desirable because of the way the bridge probes work in the experiment, as shown below. </p>
<p><p align="center">
  <img src="https://github.com/user-attachments/assets/e452ca8d-cf7d-4e64-8203-737079aaa11c" width="600" />
  <br />
  <span> The 3 probe system.</span>
</p></p>
<p>Upon every mRNA transcript of interest, several padlock-probes are attached. These stay fixed in place throughout the experiment. To illuminate each gene with the expected dye in each round, bridge probes (which are gene-specific on one arm and dye-specific on the other) are transported to all the padlock probes associated with this spot and ligated.</p>
<p>The brightness of each gene in each round is proportional to the amount of bridge probes that have ligated. Unfortunately, the number of bridge probes varies wildly between genes and rounds, giving rise to systematic differences in brightness between genes and rounds. </p>
<p>Normalising the brightnesses of each spot in each round is a way to get around this problem.</p>
<p><p align="center">
  <img src="https://github.com/user-attachments/assets/be170d2d-9dc1-4bb1-bfaa-eaca90a3d0a8" width="600" />
  <br />
  <span> Spots assigned to the gene Chrm 1 had many more bridge probes in round 4 than round 3, leading to systematic brightness differences.</span>
</p></p>
<p>Property 2 is only approximately true, but independence between rounds makes the formula for the probability of a spot being assigned to a gene much simpler.</p>
</details>
<p>Let:</p>
<ul>
<li>
<p><span class="arithmatex">\(\mathbf{f} = (\mathbf{f_1}, \ldots, \mathbf{f_{n_r}}) ^ T\)</span> be the <span class="arithmatex">\((n_r \times n_c)\)</span> round-normalised fluorescence matrix of the spot,</p>
</li>
<li>
<p><span class="arithmatex">\(\mathbf{b}_g = (\mathbf{B_{d(g, 1)}}, \ldots, \mathbf{B_{d(g, n_r)}}) ^ T\)</span> be the <span class="arithmatex">\((n_r \times n_c)\)</span> round-normalised bled code for gene <span class="arithmatex">\(g\)</span>.</p>
</li>
</ul>
<p>We define the probability of spot <span class="arithmatex">\(s\)</span> being assigned to gene <span class="arithmatex">\(g\)</span> as</p>
<div class="arithmatex">\[ \mathbb{P}[G = g \mid \mathbf{F} = \mathbf{f}] = \frac{\exp(\kappa \mathbf{b}_g \cdot \mathbf{f})}{\sum_{g'} \exp(\kappa\mathbf{b}_{g'} \cdot \mathbf{f})},\]</div>
<p>where <span class="arithmatex">\(\kappa\)</span> is a concentration parameter which controls how much the probabilities are spread out among the genes.</p>
<div class="admonition tip">
<p class="admonition-title">How to choose <span class="arithmatex">\(\kappa\)</span>?</p>
<p>The parameter <span class="arithmatex">\(\kappa\)</span> is set by adjusting the config parameter <code>kappa</code> and has default value 2. The value of <span class="arithmatex">\(\kappa\)</span> controls how much the probabilities are spread out among the genes, but does not influence the gene ordering.</p>
<ul>
<li>
<p><span class="arithmatex">\(\kappa = 0\)</span> yields a uniform distribution of probabilities between all genes,</p>
</li>
<li>
<p><span class="arithmatex">\(\kappa \rightarrow \infty\)</span> yields a distribution that tends to 1 for the gene with the maximum dot product and 0 for all others.</p>
</li>
</ul>
<p><p align="center">
  <img src="https://github.com/user-attachments/assets/1171503d-f64d-4a82-b675-9ef8bb432cf4" width="600" />
  <br />
  <span> Effects of varying <span class="arithmatex">\(\kappa\)</span> on the probabilities of a single spot.</span>
</p></p>
<p>When working with larger gene panels, all probabilities are spread out more naturally, so it helps to increase <span class="arithmatex">\(\kappa\)</span> so that probabilities have a consistent interpretation. Out current implementation sets <span class="arithmatex">\(\kappa = 2\)</span> if <span class="arithmatex">\(n_g &lt; 200\)</span> and 3 otherwise.</p>
</div>
<details class="info">
<summary>Gene Probability Derivation</summary>
<h4 id="dye-probabilities">Dye Probabilities</h4>
<p>We'll model the normalised round fluorescence vectors <span class="arithmatex">\(\mathbf{F_r}\)</span> arising from dye <span class="arithmatex">\(d\)</span> as being random and distributed according to a von Mises-Fisher distribution with mean <span class="arithmatex">\(\mathbf{B_d}\)</span> and concentration parameter <span class="arithmatex">\(\kappa\)</span>.</p>
<p>This model has probability density function</p>
<div class="arithmatex">\[\mathbb{P}[\mathbf{F_{r}} = \mathbf{f_r} \mid D = d] =  M_{\kappa} \exp(\kappa\mathbf{f_r} \cdot \mathbf{B_d}),\]</div>
<p>where <span class="arithmatex">\(\mathbf{f_r}\)</span> is a unit vector and <span class="arithmatex">\(M_{\kappa}\)</span> is a normalization constant we donâ€™t need to worry about.</p>
<h4 id="gene-probabilities">Gene Probabilities</h4>
<p>Now let <span class="arithmatex">\(\mathbf{F} = (\mathbf{F_1}, \ldots, \mathbf{F_{n_r}}) ^ T\)</span> be the <span class="arithmatex">\((n_r \times n_c)\)</span> matrix of normalised fluorescence vectors of each round <span class="arithmatex">\(r\)</span> of a spot <span class="arithmatex">\(s\)</span>. By independence between rounds, the probability of observing the fluorescence <span class="arithmatex">\(\mathbf{f}\)</span> from a spot of gene <span class="arithmatex">\(g\)</span> is just the product of the probabilities that each round <span class="arithmatex">\(r\)</span> is assigned to dye <span class="arithmatex">\(d(g, r)\)</span>. In equations, this simplifies nicely to:</p>
<div class="arithmatex">\[ 
\begin{aligned}
\mathbb{P}[\mathbf{F} = \mathbf{f} \mid G = g] &amp;= \prod_r \mathbb{P}[\mathbf{F_{r}} = \mathbf{f_r} \mid D = d(g, r)] \\ 
&amp;= \prod_r M_{\kappa} \exp \left( \kappa\mathbf{f_r} \cdot \mathbf{B_{d(g, r)}} \right) \\
&amp;= M_{\kappa}^{n_r} \exp \left( \kappa \sum_r \mathbf{f_r} \cdot \mathbf{B_{d(g, r)}} \right) \\ 
&amp;=  M_{\kappa}^{n_r} \exp(\kappa \mathbf{f \cdot b_g}), \end{aligned}
\]</div>
<p>where</p>
<ul>
<li>
<p><span class="arithmatex">\(\mathbf{f} = (\mathbf{f_1}, \ldots, \mathbf{f_{n_r}}) ^ T\)</span> is the observed round-normalised <span class="arithmatex">\((n_r \times n_c)\)</span> fluorescence matrix of the spot,</p>
</li>
<li>
<p><span class="arithmatex">\(\mathbf{b_g} = (\mathbf{B_{d(g, 1)}}, \ldots, \mathbf{B_{d(g, n_r)}}) ^ T\)</span> is the <span class="arithmatex">\((n_r \times n_c)\)</span> matrix of the bled code for gene <span class="arithmatex">\(g\)</span>,</p>
</li>
<li>
<p>the dot product <span class="arithmatex">\(\mathbf{f \cdot b_g}\)</span> is the <a href="https://en.wikipedia.org/wiki/Frobenius_inner_product">Frobenius Inner Product for Matrices</a>, ie: the sum of the elementwise product of the two matrices.</p>
</li>
</ul>
<p>We have so far only defined the probability of <span class="arithmatex">\(\mathbf{F} = \mathbf{f}\)</span> given <span class="arithmatex">\(G = g\)</span>. We can find the probability of <span class="arithmatex">\(G = g\)</span> given <span class="arithmatex">\(\mathbf{F} = \mathbf{f}\)</span> using Bayes' Rule:</p>
<div class="arithmatex">\[ 
\mathbb{P}[G = g \mid \mathbf{F} = \mathbf{f}] 
= \dfrac{\mathbb{P}[\mathbf{F} = \mathbf{f} \mid G = g] \mathbb{P}[G = g]}{ \mathbb{P}[\mathbf{F} = \mathbf{f}]}. 
\]</div>
<p>For the priors, we will assume that:</p>
<ul>
<li>
<p><span class="arithmatex">\(\mathbb{P}[G = g] = \frac{1}{n_g}\)</span> (ie: all genes are equally likely),</p>
</li>
<li>
<p><span class="arithmatex">\(\mathbb{P}[\mathbf{F} = \mathbf{f}] = \sum_g \mathbb{P}[\mathbf{F} = \mathbf{f} \mid G = g] \mathbb{P}[G = g] = \frac{1}{n_g} \sum_g M_{\kappa}^{n_r} \exp(\kappa \mathbf{b}_g \cdot \mathbf{f})\)</span>, (ie: <span class="arithmatex">\(\mathbf{f}\)</span> comes from one of the genes)</p>
</li>
</ul>
<p>This gives us the final probability:</p>
<div class="arithmatex">\[ \mathbb{P}[G = g \mid \mathbf{F} = \mathbf{f}] = \frac{\exp(\kappa \mathbf{b}_g \cdot \mathbf{f})}{\sum_g \exp(\kappa\mathbf{b}_g \cdot \mathbf{f} )}\]</div>
</details>
<h3 id="2-bleed-matrix-calculation">2: Bleed Matrix Calculation</h3>
<p>The purpose of this step is to compute an updated estimate of the bleed matrix. </p>
<p>Set some probability threshold <span class="arithmatex">\(\gamma\)</span> (in the config file <span class="arithmatex">\(\gamma\)</span> is called <code>gene_prob_thresold</code> and has default value 0.9). We define the following sets:</p>
<div class="arithmatex">\[ 
\mathcal{S} = \{ s : p(s) \geq \gamma \},
\]</div>
<div class="arithmatex">\[ 
G_{rd} = \{ g : d(g,r) = d \},
\]</div>
<div class="arithmatex">\[ 
J_{rd} = \{ \mathbf{F_{sr}} \in \mathbb{R}^{n_c}: s \in \mathcal{S}, \ g_s \in G_{rd} \}
\]</div>
<p>In words, these can be described as follows:</p>
<ul>
<li>
<p><span class="arithmatex">\(\mathcal{S}\)</span> is the set of spots with <span class="arithmatex">\(s\)</span> with probability <span class="arithmatex">\(p(s) &gt; \gamma\)</span>, ie: the high probability spots,</p>
</li>
<li>
<p><span class="arithmatex">\(G_{rd}\)</span> is the set of genes with dye <span class="arithmatex">\(d\)</span> in round <span class="arithmatex">\(r\)</span>,</p>
</li>
<li>
<p><span class="arithmatex">\(J_{rd}\)</span> is the set of colours of high probability spots assigned to genes with dye <span class="arithmatex">\(d\)</span> in round <span class="arithmatex">\(r\)</span>.</p>
</li>
</ul>
<p>By taking the union of <span class="arithmatex">\(J_{rd}\)</span> across rounds, we end up with a set of reliable colour vector estimates for dye <span class="arithmatex">\(d\)</span>:</p>
<div class="arithmatex">\[ 
\mathcal{J}_d = \bigcup_r J_{rd}
\]</div>
<details class="info">
<summary>Why do we find spots like this?</summary>
<p>The simpler way to find represantitive colours for each dye would be to look at the colours for all spots <span class="arithmatex">\(s\)</span> where <span class="arithmatex">\(\mathbf{F_{sr}} \cdot \mathbf{B_{raw, d}}\)</span> is above some threshold. This would give us a set of colours which are likely to be from dye <span class="arithmatex">\(d\)</span>. </p>
<p>Our method is better for two reasons:</p>
<ol>
<li>The raw bleed matrix <span class="arithmatex">\(\mathbf{B_{raw}}\)</span> is not always good estimate of the bleed matrix.</li>
<li>The central dyes have very similar colour spectra, so it is difficult to classify which dye the vector comes from by looking at round <span class="arithmatex">\(r\)</span> alone. By using information from adjacent rounds, we can more confidently ensure that the colours we are looking at are from dye <span class="arithmatex">\(d\)</span>.</li>
</ol>
</details>
<p>Let <span class="arithmatex">\(\mathbf{J}\)</span> be the <span class="arithmatex">\((n_{\text{good spots}} \times n_c)\)</span> matrix form of the set <span class="arithmatex">\(\mathcal{J}_d\)</span>. This just means each row of <span class="arithmatex">\(\mathbf{J}\)</span> corresponds to a good spot and each column corresponds to a channel. Compute the first <a href="https://en.wikipedia.org/wiki/Singular_value_decomposition">singular vectors</a> of <span class="arithmatex">\(\mathbf{J}\)</span>, ie: the optimal unit vectors <span class="arithmatex">\(\boldsymbol{\omega} \in \mathbb{R}^{n_c}\)</span> and <span class="arithmatex">\(\boldsymbol{\eta} \in \mathbb{R}^{n_{\text{good spots}}}\)</span> such that</p>
<div class="arithmatex">\[ J_{s, c} \approx \lambda \eta_s \omega_c, \]</div>
<p>for some scalar <span class="arithmatex">\(\lambda\)</span>. We then set <span class="arithmatex">\(\mathbf{B_d} = \boldsymbol{\omega}\)</span>, which is a normalised fluorescence vector for dye <span class="arithmatex">\(d.\)</span></p>
<h3 id="3-free-bled-code-estimation">3: Free Bled Code Estimation</h3>
<p>The purpose of this step is to estimate a representative colour, which we call a <em><strong>free bled code</strong></em> <span class="arithmatex">\(E_{grc}\)</span> for each gene <span class="arithmatex">\(g\)</span>. </p>
<div class="admonition question">
<p class="admonition-title">What makes these codes free?</p>
<p><span class="arithmatex">\(E_{grc}\)</span> is <em>free</em> in the sense that for each gene <span class="arithmatex">\(g\)</span>, <span class="arithmatex">\(E_{grc}\)</span> is only determined by spots assigned to gene <span class="arithmatex">\(g\)</span> and not by spots assigned to other genes.</p>
</div>
<p>Our method of estimating the tile-independent free bled codes <span class="arithmatex">\(\mathbf{E_{g}}\)</span> (and similarly the tile-dependent free bled codes <span class="arithmatex">\(\mathbf{D_{g,t}}\)</span>) should satisfy the following properties:</p>
<ul>
<li>
<p>If we have no spots, we should use a prior vector <span class="arithmatex">\(\mathbf{E_g} = (\mathbf{B_{d(g, 1)}}, \ldots, \mathbf{B_{d(g, n_r)}}) ^ T\)</span>,</p>
</li>
<li>
<p>We should allow each round <span class="arithmatex">\(\mathbf{E_{gr}}\)</span> to scale <span class="arithmatex">\(\mathbf{B_{d(g, r)}}\)</span> easily,</p>
</li>
<li>
<p>We should allow each round <span class="arithmatex">\(\mathbf{E_{gr}}\)</span> to change the direction of <span class="arithmatex">\(\mathbf{B_{d(g, r)}}\)</span> less easily, but still allow it to change.</p>
</li>
</ul>
<details class="note">
<summary>Why these properties?</summary>
<ul>
<li>
<p>Property 1 is necessary because for large gene panels we often have very few reads of each gene, meaning that we have very few samples to compute <span class="arithmatex">\(\mathbf{E_g}\)</span> and even fewer to compute <span class="arithmatex">\(\mathbf{D_{g, t}}\)</span>.</p>
</li>
<li>
<p>Property 2 is necessary because, as mentioned previously, different concentrations of bridge probes lead to systematic differences in brightness between genes in different rounds. We want to allow the brightness of each gene in each round to be scaled up or down without needing very many samples to do so.</p>
</li>
<li>
<p>Property 3 is necessary because sometimes the way that a particular dye is expressed varies from gene to gene. An example of this is when the dyes are not completely washed out between rounds, leading to a small amount of bleedthrough from the previous round. </p>
</li>
</ul>
<p>In the example below, both CPLX2 and FOS have dye 2 in their codes (R5 and R4 respectively), but due to incomplete washout of dye 0 in R3 of CPLX2 these genes have very different codes for dye 2.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">CPLX2</label><label for="__tabbed_1_2">FOS</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p><p align="center">
  <img src="https://github.com/user-attachments/assets/f94d2b84-bc69-4855-9601-b2bc12733866" width="600" />
  <br />
  <span> Bleedthrough into R5, Dye 2.</span>
</p></p>
</div>
<div class="tabbed-block">
<p><p align="center">
  <img src="https://github.com/user-attachments/assets/9f9ae28b-ec86-4318-a123-f78621087066" width="600" />
  <br />
  <span> No bleedthrough into R6, D2.</span>
</p></p>
</div>
</div>
</div>
</details>
<p>The following mean satisfies all the properties mentioned above. Given <span class="arithmatex">\(n\)</span> round fluorescence vectors <span class="arithmatex">\(\mathbf{f_1}, \ldots, \mathbf{f_n}\)</span> and a prior unit vector <span class="arithmatex">\(\mathbf{b}\)</span>, all in <span class="arithmatex">\(\mathbb{R}^{n_c}\)</span>, we define the <em>parallel bayes mean</em> of these as </p>
<div class="arithmatex">\[
\mathbf{\bar{F}}_{\alpha\beta}(\mathbf{b}) = \dfrac{\alpha^2}{n + \alpha^2} \mathbf{b} + \dfrac{1}{n + \alpha^2} \bigg( \sum_i \mathbf{f_i \cdot b} \bigg) \mathbf{b} + 
\dfrac{1}{n+\beta^2} \sum_i \bigg( \mathbf{f_i} - (\mathbf{f_i} \cdot \mathbf{b})\mathbf{b} \bigg).
\]</div>
<p>The values <span class="arithmatex">\(\alpha^2\)</span> and <span class="arithmatex">\(\beta^2\)</span> are in the config file as <code>concentration_parameter_parallel</code> (default value 10) and <code>concentration_parameter_perpendicular</code> (default value 50) respectively.</p>
<details class="tip">
<summary>How to choose and interpret <span class="arithmatex">\(\alpha\)</span> and <span class="arithmatex">\(\beta\)</span>?</summary>
<p>The formula for <span class="arithmatex">\(\mathbf{\bar{F}}_{\alpha \beta}(\mathbf{b})\)</span> is quite complcated, but it is actually quite easy to interpret once we understand what it is doing for different values of <span class="arithmatex">\(\alpha\)</span> and <span class="arithmatex">\(\beta\)</span>.</p>
<ul>
<li>
<p>If <span class="arithmatex">\(\alpha = \beta = 0\)</span>, this is just the average. ie: <span class="arithmatex">\(\mathbf{\bar{F}}_{0,0}(\mathbf{b}) = \frac{1}{n} \sum_i \mathbf{f_i}\)</span>,</p>
</li>
<li>
<p>if <span class="arithmatex">\(\alpha = \beta = m\)</span>, for some positive integer <span class="arithmatex">\(m\)</span>, this is the average of <span class="arithmatex">\(\mathbf{f_1}, \ldots, \mathbf{f_n}\)</span> and <span class="arithmatex">\(m\)</span> copies of <span class="arithmatex">\(\mathbf{b}\)</span>, ie: $<span class="arithmatex">\(\mathbf{\bar{F}}_{m,m}(\mathbf{b}) = \frac{1}{n + m} \sum_i \mathbf{f_i} + \frac{m}{n + m} \mathbf{b}\)</span>.</p>
</li>
<li>
<p>From the previous point, we see that if <span class="arithmatex">\(\alpha = \beta = \infty\)</span>, this is just the prior vector <span class="arithmatex">\(\mathbf{b}\)</span>, ie: <span class="arithmatex">\(\mathbf{\bar{F}}_{\infty, \infty}(\mathbf{b}) = \mathbf{b}\)</span>.</p>
</li>
<li>
<p>The component of <span class="arithmatex">\(\mathbf{\bar{F}}_{\alpha \beta}(\mathbf{b})\)</span> parallel to <span class="arithmatex">\(\mathbf{b}\)</span> has magnitude <span class="arithmatex">\(\frac{\alpha^2 + \sum_i \mathbf{f_i} \cdot \mathbf{b}}{n + \alpha^2}\)</span>, which means that:</p>
<ul>
<li>
<p>If <span class="arithmatex">\(n &lt;&lt; \alpha^2\)</span> this magnitude is approximately 1, so this component is approximately <span class="arithmatex">\(\mathbf{b}\)</span>,</p>
</li>
<li>
<p>If <span class="arithmatex">\(n &gt;&gt; \alpha^2\)</span> this magnitude is approximately <span class="arithmatex">\(\frac{1}{n} \sum_i (\mathbf{f_i} \cdot \mathbf{b} ) \mathbf{b}\)</span>, which is the magnitude of the sample mean <span class="arithmatex">\(\mathbf{\bar{f}}\)</span> in the direction <span class="arithmatex">\(\mathbf{b}\)</span>.</p>
</li>
</ul>
</li>
<li>
<p>The component of <span class="arithmatex">\(\mathbf{\bar{F}}_{\alpha \beta}(\mathbf{b})\)</span> perpendicular to <span class="arithmatex">\(\mathbf{b}\)</span> has magnitude <span class="arithmatex">\(\frac{1}{n + \beta^2} \sum_i ( \mathbf{f_i} - (\mathbf{f_i} \cdot \mathbf{b})\mathbf{b} )\)</span> which means that:</p>
<ul>
<li>
<p>If <span class="arithmatex">\(n &lt;&lt; \beta^2\)</span> this magnitude is approximately 0, so this component is approximately <span class="arithmatex">\(\mathbf{0}\)</span>,</p>
</li>
<li>
<p>If <span class="arithmatex">\(n &gt;&gt; \beta^2\)</span> this magnitude is approximately <span class="arithmatex">\(\frac{1}{n} \sum_i \mathbf{f_i} - (\mathbf{f_i} \cdot \mathbf{b})\mathbf{b}\)</span>, which is the sample mean <span class="arithmatex">\(\mathbf{\bar{f}}\)</span> perpendicular to <span class="arithmatex">\(\mathbf{b}\)</span>.</p>
</li>
</ul>
</li>
</ul>
<p>From this analysis, we see that <span class="arithmatex">\(\alpha^2\)</span> is roughly the number of spots needed to scale <span class="arithmatex">\(\mathbf{\bar{F}}_{\alpha \beta}(\mathbf{b})\)</span> in the direction of <span class="arithmatex">\(\mathbf{b}\)</span>, and <span class="arithmatex">\(\beta^2\)</span> is roughly the number of spots needed to scale <span class="arithmatex">\(\mathbf{\bar{F}}_{\alpha \beta}(\mathbf{b})\)</span> perpendicular to <span class="arithmatex">\(\mathbf{b}\)</span>. </p>
<p>This is why we to set <span class="arithmatex">\(\alpha^2 &lt;&lt; \beta^2\)</span>. We want to easily scale the average in the direction of the prior vector, but not easily change its direction.</p>
</details>
<p>We use these to estimate the free bled codes <span class="arithmatex">\(\mathbf{E_{gr}}\)</span> for each gene <span class="arithmatex">\(g\)</span> and round <span class="arithmatex">\(r\)</span> as follows:</p>
<p>Let <span class="arithmatex">\(\mathbf{f_1}, \ldots, \mathbf{f_n} \in \mathbb{R}^{n_c}\)</span> be the round <span class="arithmatex">\(r\)</span> fluorescence vectors of spots assigned to gene <span class="arithmatex">\(g\)</span> with probability greater than <span class="arithmatex">\(\gamma\)</span> and let <span class="arithmatex">\(\mathbf{B_{d(g, r)}}\)</span> be the prior unit vector. We then set each round <span class="arithmatex">\(r\)</span> to have free bled codes <span class="arithmatex">\(\mathbf{E_{gr}}\)</span> given by</p>
<div class="arithmatex">\[
\mathbf{E_{gr}} = \mathbf{\bar{F}}_{\alpha \beta}(\mathbf{B_{d(g, r)}}).
\]</div>
<p>The case for <span class="arithmatex">\(\mathbf{D_{g, t}}\)</span> is exactly analogous, except we use the fluorescence vectors of spots assigned to gene <span class="arithmatex">\(g\)</span> in tile <span class="arithmatex">\(t\)</span> with probability greater than <span class="arithmatex">\(\gamma\)</span>.</p>
<details class="info">
<summary>Derivation of the Parallel Bayes Mean</summary>
<p>The formula for the parallel bayes mean is a <a href="https://en.wikipedia.org/wiki/Maximum_a_posteriori_estimation">maximum a posteriori estimate</a>. This means that we view the data as coming from a particular distribution with some unkown mean <span class="arithmatex">\(\boldsymbol{\mu}\)</span>, which we want to estimate. We have some prior beliefs about what <span class="arithmatex">\(\boldsymbol{\mu}\)</span> should be and how this should vary, which we encode in a prior distribution of potential values for <span class="arithmatex">\(\boldsymbol{\mu}\)</span>. The observed data has a certain probability given <span class="arithmatex">\(\boldsymbol{\mu}\)</span>, and by Bayes rule each <span class="arithmatex">\(\boldsymbol{\mu}\)</span> has a probability given the data. The maximum a posteriori estimate <span class="arithmatex">\(\boldsymbol{\hat{\mu}}\)</span> is the value of <span class="arithmatex">\(\boldsymbol{\mu}\)</span> which maximises this conditional probability distribution.</p>
<p>Let <span class="arithmatex">\(\mathbf{F_1}, \ldots, \mathbf{F_n}\)</span>  be the round <span class="arithmatex">\(r\)</span> fluorescence vectors of spots assigned to gene <span class="arithmatex">\(g\)</span> with high probability and let <span class="arithmatex">\(\mathbf{B}_{d(g,r)}\)</span> be the prior unit vector.</p>
<p>To begin, assume the vectors <span class="arithmatex">\(\mathbf{F_1}, \ldots, \mathbf{F_n}\)</span> are i.i.d normal random variables with mean <span class="arithmatex">\(\boldsymbol{\mu}\)</span> and covariance <span class="arithmatex">\(I_{n_c}\)</span>, wihch means the sample mean is also normal with mean <span class="arithmatex">\(\boldsymbol{\mu}\)</span> and covariance <span class="arithmatex">\(\frac{I_{n_c}}{n}\)</span>. Impose a normal prior on the space of possible means:</p>
<div class="arithmatex">\[
\overline{\mathbf{F}} \sim \mathcal{N} \bigg( \boldsymbol{\mu}, \frac{\boldsymbol{I_{n_c}}}{n} \bigg)
\]</div>
<div class="arithmatex">\[
\boldsymbol{\mu} \sim \mathcal{N}(\mathbf{B}_{d(g,r)}, \Sigma)
\]</div>
<p>where </p>
<div class="arithmatex">\[
\Sigma = \text{Diag}\left(\frac{1}{\alpha^2}, \frac{1}{\beta^2}, \ldots, \frac{1}{\beta^2}\right),
\]</div>
<p>in the orthonormal basis <span class="arithmatex">\(\mathbf{v}_1 = \mathbf{B}_{d(g,r)}\)</span>, and everything else orthogonal to this.</p>
<p>Set <span class="arithmatex">\(\boldsymbol{\Lambda} =\boldsymbol{\Sigma}^{-1}\)</span>, <span class="arithmatex">\(\mathbf{b} = \mathbf{B_{d(g,r)}}\)</span>, and recall that the normal is a conjugate prior, meaning the posterior <span class="arithmatex">\(\boldsymbol{\mu} \mid \mathbf{\overline{F}}\)</span> is also normal. </p>
<p>To find its mode we'll solve for the zeros of the derivative of its log-density. The log-density of <span class="arithmatex">\(\boldsymbol{\mu} \mid \mathbf{\overline{F}}\)</span> is given by</p>
<div class="arithmatex">\[
\begin{aligned}
l(\boldsymbol{\mu}) &amp;= \log P(\boldsymbol{\mu}| \overline{\mathbf{F}} = \mathbf{f}) \\ \\
       &amp;= \log P(\boldsymbol{\mu}) + \log P(\overline{\mathbf{F}} = \mathbf{f} | \boldsymbol{\mu}) + C \\ \\
       &amp;= -\frac{1}{2} (\boldsymbol{\mu} - \mathbf{b})^T \boldsymbol{\Lambda} (\boldsymbol{\mu} - \mathbf{b}) - \frac{n}{2} (\boldsymbol{\mu} - \mathbf{f})^T (\boldsymbol{\mu} - \mathbf{f}) + C
\end{aligned}
\]</div>
<p>This has derivative</p>
<div class="arithmatex">\[
\frac{\partial l}{\partial \boldsymbol{\mu}} = - \boldsymbol{\Lambda} (\boldsymbol{\mu} - \boldsymbol{b}) - n(\boldsymbol{\mu} - \mathbf{f})
\]</div>
<p>Setting this to <span class="arithmatex">\(\mathbf{0}\)</span>, rearranging for <span class="arithmatex">\(\boldsymbol{\mu}\)</span> and using the fact that</p>
<div class="arithmatex">\[
\boldsymbol{\Lambda} \mathbf{v} = 
\begin{cases}
\alpha^2 \mathbf{v} &amp; \text{if } \mathbf{v} = \lambda\mathbf{b} \\ \\
\beta^2 \mathbf{v} &amp; \text{otherwise}
\end{cases}
\]</div>
<p>we get</p>
<div class="arithmatex">\[
\begin{aligned}
\boldsymbol{\hat{\mu}} &amp;= (\Lambda + nI)^{-1}(\Lambda \mathbf{b} + n\mathbf{f}) \\ \\ 
    &amp;= (\Lambda + nI)^{-1}(\alpha^2 \mathbf{b} + n\mathbf{f}) \\ \\
    &amp;= (\Lambda + nI)^{-1}(\alpha^2 \mathbf{b} + n(\mathbf{f} \cdot \mathbf{b})\mathbf{b} + n(\mathbf{f} - (\mathbf{f} \cdot \mathbf{b})\mathbf{b})) \\ \\
    &amp;= (\Lambda + nI)^{-1}((\alpha^2 + n \mathbf{f} \cdot \mathbf{b})\mathbf{b} + n(\mathbf{f} - (\mathbf{f} \cdot \mathbf{b})\mathbf{b}))\\ \\
&amp;= \dfrac{(\alpha^2 + n \mathbf{f} \cdot \mathbf{b})}{n + \alpha^2} \mathbf{b} +
 \dfrac{n}{n+\beta^2} \bigg( \mathbf{f} - (\mathbf{f} \cdot \mathbf{b})\mathbf{b} \bigg)
\end{aligned}
\]</div>
<p>Plugging in the observed sample mean <span class="arithmatex">\(\mathbf{f} = \frac{1}{n}\sum_i \mathbf{f_{i, r}}\)</span> yields our estimate <span class="arithmatex">\(\boldsymbol{\hat{\mu}}\)</span>.</p>
</details>
<h3 id="4-round-and-channel-normalisation">4: Round and Channel Normalisation</h3>
<p>The purpose of this step is to find a scale factor <span class="arithmatex">\(V_{rc}\)</span> for each round <span class="arithmatex">\(r\)</span> and channel <span class="arithmatex">\(c\)</span> which gets as many of our spots as close as possible to their target values. We will then multiply <span class="arithmatex">\(V_{rc}\)</span> by the free bled codes <span class="arithmatex">\(E_{grc}\)</span> to get the <em><strong>constrained bled codes</strong></em> <span class="arithmatex">\(K_{grc}\)</span>. </p>
<div class="admonition question">
<p class="admonition-title">What makes these codes constrained?</p>
<p>The codes <span class="arithmatex">\(K_{grc}\)</span> are <em>constrained</em> in the sense that the value of <span class="arithmatex">\(K_{grc}\)</span> is determined by several genes other than <span class="arithmatex">\(g\)</span>.</p>
<p>These codes have nice global properties, like as many genes as possible being as close as possible to their target values, but will not be representative of the spots assigned to gene <span class="arithmatex">\(g\)</span>. This is addressed in section 5, where we will be to find a scale to get the spots as close as possible to these new constrained bled codes.</p>
</div>
<p>The target values work as follows: </p>
<ul>
<li>
<p><span class="arithmatex">\(T\)</span> is defined as <code>target_values</code> in the config file as a list of length <span class="arithmatex">\(n_c\)</span>. </p>
</li>
<li>
<p><span class="arithmatex">\(T_c\)</span> is the target value for channel <span class="arithmatex">\(c\)</span> in its representative dye <span class="arithmatex">\(d_{\textrm{max}}(c)\)</span>,</p>
</li>
<li>
<p><span class="arithmatex">\(d_{\textrm{max}}\)</span> is defined as <code>d_max</code> in the config file as a list of length <span class="arithmatex">\(n_c\)</span>. </p>
</li>
<li>
<p><span class="arithmatex">\(d_{\textrm{max}}(c)\)</span> is the dye we use to represent channel <span class="arithmatex">\(c\)</span>, and we want to get its brightness in channel <span class="arithmatex">\(c\)</span>, <span class="arithmatex">\(B_{d_{\textrm{max}}(c), c}\)</span> as close as possible to <span class="arithmatex">\(T_c\)</span>.</p>
</li>
</ul>
<p>Any gene that has dye <span class="arithmatex">\(d_{\textrm{max}}(c)\)</span> in round <span class="arithmatex">\(r\)</span> will have its free bled code <span class="arithmatex">\(E_{grc}\)</span> scaled by <span class="arithmatex">\(V_{rc}\)</span> to get as close as possible to <span class="arithmatex">\(T_c\)</span>. Since <span class="arithmatex">\(E_{grc}\)</span> is a representative colour for all spots assigned to gene <span class="arithmatex">\(g\)</span>, this will also get the spots as close as possible to their target values. </p>
<p>As in section 2 above, let <span class="arithmatex">\(G_{rd}\)</span> be the set of genes with dye <span class="arithmatex">\(d\)</span> in round <span class="arithmatex">\(r\)</span> and define the loss function </p>
<div class="arithmatex">\[
L(V_{rc}) = \sum_{g \in G_{r, \ d_{max}(c)}} \sqrt{N_{g}} \  \bigg( V_{rc} \ E_{grc} - T_{c} \bigg)^2,
\]</div>
<p>where <span class="arithmatex">\(N_g\)</span> is the number of high probability spots assigned to gene <span class="arithmatex">\(g\)</span>. There is no reason this has to be a square root, though if it is not, too much influence is given to the most frequent genes. We minimise this loss to obtain the optimal value</p>
<div class="arithmatex">\[
V_{rc} = \dfrac{ \sum_{g \in G_{r, \ d_{max}(c) }} \sqrt{N_g} E_{grc} T_{c} } { \sum_{g \in G_{r, \ d_{max}(c) }} \sqrt{N_g} E_{grc}^2 },
\]</div>
<p>Now define the <em>constrained bled codes</em>, which we will just call <em>bled codes</em> </p>
<div class="arithmatex">\[
K_{grc} = E_{grc}V_{rc}.
\]</div>
<h3 id="5-tile-normalisation">5: Tile Normalisation</h3>
<p>The purpose of this step is to remove brightness differences between images from different tiles in the same round and channel. We do this by finding a scale factor <span class="arithmatex">\(Q_{trc}\)</span> for each tile <span class="arithmatex">\(t\)</span>, round <span class="arithmatex">\(r\)</span> and channel <span class="arithmatex">\(c\)</span> which gets as many of our spots on tile <span class="arithmatex">\(t\)</span> as close as possible to <span class="arithmatex">\(K_{grc}\)</span>.</p>
<p>Our method works almost identically to step 4. Let <span class="arithmatex">\(G_{rd}\)</span> be the genes with dye <span class="arithmatex">\(d\)</span> in round <span class="arithmatex">\(r\)</span>. Define the loss</p>
<div class="arithmatex">\[
L(Q_{trc}) = \sum_{g \in G_{r, \ d_{max}(c)}} \sqrt{N_{g,t}} \  \bigg( Q_{trc} \ D_{gtrc} - K_{grc} \bigg)^2,
\]</div>
<p>where <span class="arithmatex">\(N_{g, t}\)</span> is the number of high probability spots of gene <span class="arithmatex">\(g\)</span> in tile <span class="arithmatex">\(t\)</span>. </p>
<details class="note">
<summary>If Q is correcting for tile differences, why does it have indices for <span class="arithmatex">\(r\)</span> and <span class="arithmatex">\(c\)</span>?</summary>
<p>The scale factor <span class="arithmatex">\(Q_{trc}\)</span> is defined to correct for differences in brightness between tiles, but the way that the brightness varies between tiles is completely independent for different round-channel pairs. </p>
<p>This is because the cause of brightness differences between tiles is largely random from image to image, as can be observed by looking at spots in the overlapping regions of adjacent tiles in the same round and channel.</p>
</details>
<p>We minimise this loss to obtain the optimal value</p>
<div class="arithmatex">\[
Q_{trc} = \dfrac{ \sum_{g \in G_{r, \ d_{max}(c) }} \sqrt{N_{gt}} \ K_{grc}  D_{gtrc}} { \sum_{g \in G_{r,\ d_{max}(c) }} \sqrt{N_{gt}}  D_{gtrc}^2 }.
\]</div>
<h3 id="6-and-7-application-of-scales-computation-of-final-scores-and-bleed-matrix">6 and 7: Application of Scales, Computation of Final Scores and Bleed Matrix</h3>
<p>All that is left to do is multiply the spot colours <span class="arithmatex">\(F_{src}\)</span> by the updated normalisation factor <span class="arithmatex">\(Q_{trc}\)</span> to get the final spot colours: <span class="arithmatex">\(F_{src} \mapsto Q_{trc} F_{src}\)</span>.</p>
<p>We then compute the cosine similarity between each spot colour <span class="arithmatex">\(\mathbf{F_s}\)</span> and each bled code <span class="arithmatex">\(\mathbf{K_{grc}}\)</span> to get the two variables</p>
<ul>
<li><code>dot_product_gene_no[s]</code> = <span class="arithmatex">\(\textrm{argmax}_g \bigg( \dfrac{\mathbf{F_s \cdot K_{g}}}{\| \mathbf{F_s} \| \| \mathbf{K_{g}} \|} \bigg)\)</span>,</li>
<li><code>dot_product_gene_score[s]</code> = <span class="arithmatex">\(\textrm{max}_g \bigg( \dfrac{\mathbf{F_s \cdot K_{g}}}{\| \mathbf{F_s} \| \| \mathbf{K_{g}} \|} \bigg)\)</span>.</li>
</ul>
<p>We also compute probabilities for each spot <span class="arithmatex">\(s\)</span> being assigned to gene <span class="arithmatex">\(g\)</span> as</p>
<ul>
<li><code>gene_prob[s, g]</code> = <span class="arithmatex">\(\dfrac{\exp(\kappa \mathbf{K_{g} \cdot F_s})}{\sum_{g'} \exp(\kappa \mathbf{K_{g'} \cdot F_s})}\)</span>,</li>
</ul>
<p>where <span class="arithmatex">\(\mathbf{F_s}\)</span> and <span class="arithmatex">\(\mathbf{K_{g}}\)</span> have both been round-normalised. Finally, with these updated gene assignments, we can compute the final bleed matrix <span class="arithmatex">\(\mathbf{B}\)</span> in the same way as in step 2.</p>
<h2 id="diagnostics">Diagnostics</h2>
<p>Diagnosing the quality of the gene assignments is a crucial part of the pipeline. We provide several diagnostics to help with this:</p>
<h3 id="view-scaling-and-bg-removal">View Scaling And BG Removal</h3>
<p><div class="highlight"><pre><span></span><code>from coppafish.plot.call_spots import ViewScalingAndBGRemoval
ViewScalingAndBGRemoval(nb)
</code></pre></div>
(or simply press 'N' in the main results' viewer)</p>
<p align="center">
  <img src="https://github.com/user-attachments/assets/76b7f610-6898-47dc-bc43-e6ba85cc1684" width="600" />
  <br />
  <span> Viewing the background removal and scaling of a subset of isolated spots.</span>
</p>

<p>This shows a subset of 10,000 isolated spots in descending order of amount of background. The images on the top row are spot colours, each flattened into a single row and demarcated into channels by the red vertical lines. The plots on the bottom row show the intensity of a bright spot in each round channel.</p>
<p>This plot shows us a few things:</p>
<ul>
<li>
<p>Certain channels have much higher background than others. The final column is a good check that the background has been removed.</p>
</li>
<li>
<p>Different channels have different baseline brightnesses. Check that the brightnesses in the middle and final column are to your liking and similar to the target values.</p>
</li>
<li>
<p>The final brightnesses are not all the same: this is because we imposed channel-specific target values in step 4. This is a good check that the target scaling is working as expected.</p>
</li>
</ul>
<h3 id="view-bleed-matrix">View Bleed Matrix</h3>
<div class="highlight"><pre><span></span><code>from coppafish.plot.call_spots import ViewBleedMatrix
ViewBleedMatrix(nb)
</code></pre></div>
<p>(or simply press 'B' in the main results' viewer)</p>
<p align="center">
<img src="https://github.com/user-attachments/assets/27223ed8-fac3-4d70-9440-4bf3637ef87e" width="600" />
<br />
<span> Viewing the bleed matrix.</span>
</p>

<p>This viewer shows 3 bleed matrices, each with columns (dyes) normalised.</p>
<ul>
<li>
<p>The first is the raw bleed matrix <span class="arithmatex">\(\mathbf{B_{raw}}\)</span> which is the initial estimate of the bleed matrix, used for the very first gene assignments in step 1.</p>
</li>
<li>
<p>The second is the initial bleed matrix <span class="arithmatex">\(\tilde{\mathbf{B}}\)</span> made from an SVD of high probability spots. This is scaled according to the initial scale factor <span class="arithmatex">\(\tilde{A}_{trc}\)</span> introduced in step 0. This is why channel 10 is so much brighter than its target value.</p>
</li>
<li>
<p>The final bleed matrix <span class="arithmatex">\(\mathbf{B}\)</span> is the bleed matrix estimated from the final gene assignments, on high probability spots. This is scaled according to the final scale factor <span class="arithmatex">\(A_{trc} = Q_{trc}\tilde{A}_{trc}\)</span>. You should be able to see the values are roughly in the same ratios as the target values.</p>
</li>
</ul>
<h3 id="view-free-and-constrained-bled-codes">View Free And Constrained Bled Codes</h3>
<div class="highlight"><pre><span></span><code>from coppafish.plot.call_spots import ViewFreeAndConstrainedBledCodes
ViewFreeAndConstrainedBledCodes(nb)
</code></pre></div>
<p>This will pull up a viewer that shows you the free bled codes <span class="arithmatex">\(E_{grc}\)</span> and the constrained bled codes <span class="arithmatex">\(K_{grc}\)</span> from the most influential genes for a given round and channel. This is a good way to check if the target scale <span class="arithmatex">\(V_{rc}\)</span> is working as expected.</p>
<p>To view different rounds and channels, simply scroll.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="2:2"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><div class="tabbed-labels"><label for="__tabbed_2_1">R0C5</label><label for="__tabbed_2_2">R2C15</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p><p align="center">
<img src="https://github.com/user-attachments/assets/2720d341-6513-4ad2-a9cf-e0ba71d65037" width="600" />
<br />
<p></p>
</div>
<div class="tabbed-block">
<p><p align="center">
<img src="https://github.com/user-attachments/assets/36bcf922-69ae-41b5-92a4-63625a416140" width="600" />
<br />
<p></p>
</div>
</div>
</div>
<p>If this works as expected, the constrained bled codes should have values close to their target values and the constrained bled codes should be more homogeneous than the free bled codes. This can be seen in the first image, where R0C5 is initially very bright, but after scaling is much closer to the brightnesses of the other rounds and channels.</p>
<h3 id="view-target-regression">View Target Regression</h3>
<div class="highlight"><pre><span></span><code>from coppafish.plot.call_spots import ViewTargetRegression
ViewTargetRegression(nb)
</code></pre></div>
<p>This viewer is similar to the previous one in that it is showing how well the target scaling is working. It does this in a bit more detail, but is a little confusing!</p>
<p>To view different rounds and channels, simply scroll.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="3:2"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><input id="__tabbed_3_2" name="__tabbed_3" type="radio" /><div class="tabbed-labels"><label for="__tabbed_3_1">R0C27</label><label for="__tabbed_3_2">R4C5</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p><p align="center">
<img src="https://github.com/user-attachments/assets/718a4536-b70f-4025-a9b7-1e32e449f1d7" width="600" />
<br />
<p></p>
</div>
<div class="tabbed-block">
<p><p align="center">
<img src="https://github.com/user-attachments/assets/d9df4bff-2d02-4df2-9b08-26d0d907783f" width="600" />
<br />
<p></p>
</div>
</div>
</div>
<p>In the plots above:</p>
<ul>
<li>Each dot is a gene, which has dyed <span class="arithmatex">\(d_{\textrm{max}}(c)\)</span> in round <span class="arithmatex">\(r\)</span>.</li>
<li>The size of the dot is proportional to the number of spots assigned to that gene.</li>
<li>The x-axis values are completely random.</li>
<li>In the leftmost column, the y-axis values are the brightnesses <span class="arithmatex">\(E_{grc}\)</span> of the genes after the initial scaling <span class="arithmatex">\(\tilde{A}_{trc}\)</span> but before the target scaling <span class="arithmatex">\(V_{trc}\)</span>.</li>
<li>In the middle column, the y-axis values are the brightnesses <span class="arithmatex">\(K_{grc}\)</span> of the genes after the target scaling <span class="arithmatex">\(V_{trc}\)</span>.</li>
</ul>
<p>It is important to check how well each round and channel is being scaled to its target value. R0C27 is pretty good, with most of the genes being pretty concentrated at the target value. R4C5 is much noisier, with many genes consistently too bright or too dim.</p>
<h3 id="view-tile-scale-regression">View Tile Scale Regression</h3>
<p><div class="highlight"><pre><span></span><code>from coppafish.plot.call_spots import ViewTileScaleRegression
ViewTileScaleRegression(nb, t)
</code></pre></div>
This function looks at a fixed tile and then shows the regression for the tile scale factor <span class="arithmatex">\(Q_{trc}\)</span> for each round and channel. Recall that this is the scale factor that multiplies the tile-dependent free bled codes <span class="arithmatex">\(D_{gtrc}\)</span> to get the constrained bled codes <span class="arithmatex">\(K_{grc}\)</span>.</p>
<p>As in the previous diagnostic, each spot is a gene with dye <span class="arithmatex">\(d_{\textrm{max}}(c)\)</span> in round <span class="arithmatex">\(r\)</span> and the size of the dot is proportional to the number of spots assigned to that gene. Unlike the previous diagnostic, in this plot the x-values are not random, but are the brightnesses <span class="arithmatex">\(D_{gtrc}\)</span> of the genes (averaged from spots which have been multiplied by initial scale <span class="arithmatex">\(\tilde{A}_{trc}\)</span>). The y-values are the brightnesses <span class="arithmatex">\(K_{grc}\)</span>. </p>
<p>A couple of things to note:</p>
<ul>
<li>
<p>Different slopes within the same column (channel) indicate that this is picking up on differences in round brightnesses for this channel on this tile.</p>
</li>
<li>
<p>If these regressions have a low <span class="arithmatex">\(R^2\)</span> value, the tile scaling is not working well. This may be a sign of a blank tile or poor registration.</p>
</li>
</ul>
<p align="center">
    <img src="https://github.com/user-attachments/assets/858ce3c0-b603-4633-b574-03a81ffd21c9" width="600" />
    <br />
    <span> Viewing the background removal and scaling of a subset of isolated spots.</span>
<p>

### View Scale Factors

<div class="highlight"><pre><span></span><code>from coppafish.plot.call_spots import ViewScaleFactors
ViewScaleFactors(nb)
</code></pre></div>

This simple viewer shows the target scale $V_{rc}$, the tile scale $Q_{trc}$ and the relative scale $Q_{trc}/V_{rc}$ for each round and channel. 

What to expect:

- The tile scale $Q_{trc}$ should be close to $V_{rc}$ for each tile $t$. This is because $D_{gtrc}Q_{trc} \approx K_{grc = E_{grc}V_{rc}}$. So if $E_{grc} \approx D_{gtrc}$, then $Q_{trc} \approx V_{rc}$.

- The relative scale measures how much we deviate from the vase where $E_{grc} = D_{gtrc}$, and should not have a huge amount of variance. In the plot above the highest value is 0.5 and the lowest is 0.35, which is a good range.

<p align="center">
<img src="https://github.com/user-attachments/assets/c7333f48-8a19-4129-8730-00db6723f472" width="600" />
<br />
<p>

### Gene Efficiency Viewer
<div class="highlight"><pre><span></span><code>from coppafish.plot.call_spots import GeneEfficiencyViewer
GeneEfficiencyViewer(nb, score_threshold=gamma, mode=gene_assignment_mode)
</code></pre></div>

(or simply press 'E' in the main results' viewer)

<p align="center">
<img src="https://github.com/user-attachments/assets/6d8de8b7-e385-4d74-9324-035baf053031" width="600" />
<br />
<p>

Each row represents a gene and each column a round. The colour of each cell is the amount of weight that gene $g$ has in round $r$. The ideal case would be homogeneous colours across the rows, indicating that each gene is equally bright in each round, but this is never the case.

Look out for:

- Genes with an abnormal amount of spots assigned to them. This is often the case for poor quality genes which look a lot like background. If this is the case, the gene probability threshod `gene_prob_thresh` in the config file should be increased.

- Genes with very high or low gene efficiencies. This should not happen if `concentration_parameter_parallel` is sufficiently high, as it typically should need at least 10 spots to scale each dye. If the gene efficiencies are incorrect, OMP will struggle to find the correct gene assignments.

### Gene Spots Viewer
<div class="highlight"><pre><span></span><code>from coppafish.plot.call_spots import GeneSpotsViewer
GeneSpotsViewer(nb, score_threshold=gamma, gene_index=g, mode=gene_assignment_mode)
</code></pre></div>

(or simply click one of the genes in the gene efficiency viewer)

<p align="center">
<img src="https://github.com/user-attachments/assets/a0996167-5878-45e7-a609-4269dde8736a" width="600" />
<br />
<p>

This viewer shows the spots assigned to a particular gene above a certain threshold, under a certain gene assignment mode (either 'anchor', 'prob' or 'omp'). 

This is the viewer I use the most. It helps me find abnormalities that would be hard to spot otherwise, like the persistent unexpected channel 27 signal in round 0 in the images above. If a particular gene is under or over expressed, this viewer will typically tell us why. It also gives us a very nice representation of which dyes, rounds and channels are clean and which are noisy.












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.tabs.sticky", "content.code.annotation", "content.code.copy"], "search": "../assets/javascripts/workers/search.07f07601.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.56dfad97.min.js"></script>
      
        <script src="../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="../javascripts/katex.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/katex.min.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/contrib/auto-render.min.js"></script>
      
    
  </body>
</html>